src/xen/xen.c

```diff
@@ -0,0 +1,847 @@
+    *state = val;
+    return s;
+}
+
+status xenbus_set_state(u32 tx_id, buffer path, XenbusState newstate)
+{
+    XenbusState oldstate;
+    status s = xenbus_get_state(path, &oldstate);
+    if (!is_ok(s))
+        return s;
+    xen_debug("%s: old  %d, new %d", __func__, oldstate, newstate);
+    if (oldstate == newstate)
+        return STATUS_OK;
+    return xenstore_sync_printf(tx_id, path, "state", "%d", newstate);
+}
+
+static status traverse_directory_internal(heap h, buffer path, tuple *parent)
+{
+    xenstore_debug("%s: path \"%s\"", __func__, path);
+    buffer response = allocate_buffer(h, PAGESIZE);
+
+    status s = xenstore_sync_request(0, XS_DIRECTORY, path, response);
+    if (!is_ok(s))
+        goto out; /* XXX should add context to status */
+
+    if (buffer_length(response) == 0) {
+        *parent = 0;            /* indicate zero response; leaf node */
+        goto out;
+    }
+
+    *parent = allocate_tuple();
+
+    buffer splice = allocate_buffer(h, buffer_length(path) + 16);
+    buffer_write(splice, buffer_ref(path, 0), buffer_length(path) - 1);
+    push_u8(splice, '/');
+    bytes splice_saved_end = splice->end;    /* XXX violation; amend buffer interface */
+
+    do {
+        char * child = buffer_ref(response, 0);
+        int child_len = runtime_strlen(child) + 1;
+        buffer_write(splice, child, child_len);
+
+        value child_node = 0;
+        s = traverse_directory_internal(h, splice, (tuple *)&child_node);
+        if (!is_ok(s))
+            goto out;
+
+        if (!child_node) {
+            /* leaf node, read content */
+            buffer leaf_value = allocate_buffer(h, 8);
+            assert(leaf_value != INVALID_ADDRESS); /* XXX iron out inconsistencies */
+            s = xenstore_sync_request(0, XS_READ, splice, leaf_value);
+            if (!is_ok(s)) {
+                deallocate_buffer(leaf_value);
+                goto out_slice;
+            }
+            child_node = (value)leaf_value;
+        }
+        table_set(*parent, sym_this(child), child_node);
+        buffer_consume(response, child_len);
+        splice->end = splice_saved_end; /* XXX see above */
+    } while (buffer_length(response) > 0 && *(u8 *)buffer_ref(response, 0) != '\0');
+  out_slice:
+    deallocate_buffer(splice);
+  out:
+    deallocate_buffer(response);
+    return s;
+}
+
+static inline status traverse_directory(heap h, const char * path, tuple *node)
+{
+    return traverse_directory_internal(h, alloca_wrap_buffer(path, runtime_strlen(path) + 1), node);
+}
+
+status xen_probe_devices(void)
+{
+    xen_debug("probing xen device tree from xenstored");
+    assert(xen_info.device_tree == 0);
+    tuple node = 0;
+    status s = traverse_directory(xen_info.h, "device", &node);
+    if (!is_ok(s))
+        return s;
+    if (!node)
+        return timm("result", "failed to parse directory");
+
+    xen_info.device_tree = node;
+    xen_debug("success; result: %v", node);
+
+    table_foreach(node, k, v) {
```

Comment:
If I'm reading this correctly, the leaves here are never deallocated ...

Reply:
The intention here was that the device probe routine should consume the passed tuple entirely, but I'm sure I'm not doing that. I'll clean it up before merging.

Reply:
Sadly, this depends on #974.